filesystem/fat32/directoryentry.go

```diff
@@ -413,11 +416,47 @@ func convertLfnSfn(name string) (string, string, bool, bool) {
 		isLFN = true
 	}
 
+	truncate := func(name string, n int) string {
+		return name[0:6] + "~" + strconv.Itoa(n)
+	}
+
+	hash := func(shortName string, n int) string {
+		h := fnv.New32a()
+		// the input name may have more information than the short name
+		h.Write([]byte(name))
+		return shortName[0:2] + fmt.Sprintf("%04X", (h.Sum32()%0x10000)) + "~" + strconv.Itoa(n)
+	}
+
+	random := func() string {
+		return fmt.Sprintf("%06X", (rand.Int63()%0x1000000)) + "~" + strconv.Itoa(1+rand.Intn(8))
+	}
+
 	// convert shortName to 8 chars
 	if len(shortName) > 8 {
 		isLFN = true
 		isTruncated = true
-		shortName = shortName[:6] + "~" + "1"
+		if !cache[truncate(shortName, 9)+"."+extension] {
+			i := 1
+			for i < 9 && cache[truncate(shortName, i)+"."+extension] {
+				i++
+			}
+			shortName = truncate((shortName), i)
```

Comment:
This calls `truncate()` duplicatively several times. Can we tighten it up a bit?

Also, why is this:

```go
			i := 1
 			for i < 9 && cache[truncate(shortName, i)+"."+extension] {
 				i++
 			}
```

and not the simpler and more idiomatic:

```go
			var found string
 			for i := i; i++; i < 9 {
                            found = truncate(shortName, i)+"."+extension
                            if _, ok := cache[found]; ok {
                               break
                            }
 			}
                        shortName = found
```

or similar